site/en/docs/android/custom-tabs/howto-custom-tab-native-apps/index.md

```diff
@@ -0,0 +1,112 @@
+                    Intent.FLAG_ACTIVITY_REQUIRE_NON_BROWSER);
+    try {
+        context.startActivity(nativeAppIntent);
+        return true;
+    } catch (ActivityNotFoundException ex) {
+        return false;
+    }
+}
+```
+
+## Before Android 11
+
+Even though the application may target Android 11, or API level 30, previous Android versions will
+not understand the `FLAG_ACTIVITY_REQUIRE_NON_BROWSER` flag. To support those devices, we can query the Package Manager. Our approach is to  query the Package Manager for applications that support a generic "http" intent. Those applications are likely browsers.
+
+Then, query for applications that handle intents for the specific URL we want to launch. This will
+return both browsers and native applications setup to handle that URL.
+
+Now, remove all browsers found on the first list from the second list, and we'll be left only with
+native apps.
+
+If the list is empty, we know there are no native handlers and return false. Otherwise, we launch
+the intent for the native handler.
+
+```java
+private static boolean launchNativeBeforeApi30(Context context, Uri uri) {
+    PackageManager pm = context.getPackageManager();
+
+    // Get all Apps that resolve a generic url
+    Intent browserActivityIntent = new Intent()
+            .setAction(Intent.ACTION_VIEW)
+            .addCategory(Intent.CATEGORY_BROWSABLE)
+            .setData(Uri.fromParts("http", "", null));
+    Set<String> genericResolvedList = extractPackageNames(
+            pm.queryIntentActivities(browserActivityIntent, 0));
+
+    // Get all apps that resolve the specific Url
+    Intent specializedActivityIntent = new Intent(Intent.ACTION_VIEW, uri)
+            .addCategory(Intent.CATEGORY_BROWSABLE);
+    Set<String> resolvedSpecializedList = extractPackageNames(
+            pm.queryIntentActivities(specializedActivityIntent, 0));
+
+    // Keep only the Urls that resolve the specific, but not the generic
+    // urls.
+    resolvedSpecializedList.removeAll(genericResolvedList);
+
+    // If the list is empty, no native app handlers were found.
+    if (resolvedSpecializedList.isEmpty()) {
+        return false;
+    }
+
+    // We found native handlers. Launch the Intent.
+    specializedActivityIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+    context.startActivity(specializedActivityIntent);
+    return true;
+}
+```
+
+
+### Putting it all together
+
+We need to ensure we are using the right method for each occasion: 
+
+```java
+static void launchUri(Context context, Uri uri) {
+    boolean launched = Build.VERSION.SDK_INT >= 30 ?
+            launchNativeApi30(context, uri) :
+            launchNativeBeforeApi30(context, uri);
+
+    if (!launched) {
+        new CustomTabsIntent.Builder()
+                .build()
+                .launchUrl(context, uri);
+    }
+}
+```
+
+`Build.VERSION.SDK_INT` provides the information we need. If it's equal or larger than 30, Android
```

Comment:
```suggestion
`Build.VERSION.SDK_INT` provides the information we need. If it's greater than or equal to 30, Android
```

Reply:
You're text is grammatically wrong. It's just not how that idea is typically expressed in English.